src/React/Basic.purs

```diff
@@ -32,25 +29,57 @@ foreign import react_
 -- | constructed using the helper functions provided by the `React.Basic.DOM`
 -- | module (and re-exported here).
 react
-  :: forall props state
-   . { initialState :: state
-     , receiveProps :: props -> state -> (state -> Eff (react :: ReactFX) Unit) -> Eff (react :: ReactFX) Unit
-     , render :: props -> state -> (state -> Eff (react :: ReactFX) Unit) -> JSX
+  :: forall props state fx
+   . { displayName :: String
+     , initialState :: state
+     , receiveProps :: props -> state -> (SetState state fx) -> Eff (react :: ReactFX | fx) Unit
+     , render :: props -> state -> (SetState state fx) -> JSX
+     }
+  -> props
+  -> JSX
+react = createElement <<< reactComponent
+
+-- | Create a React component from a _specification_ of that component, without
+-- | pre-applying it to `createElement`. Use this function for more standard
+-- | JavaScript React interop.
+reactComponent
+  :: forall props state fx
+   . { displayName :: String
+     , initialState :: state
+     , receiveProps :: props -> state -> (SetState state fx) -> Eff (react :: ReactFX | fx) Unit
+     , render :: props -> state -> (SetState state fx) -> JSX
      }
   -> ReactComponent props
-react { initialState, receiveProps, render } =
-  react_
-    { initialState
+reactComponent { displayName, initialState, receiveProps, render } =
+  component_
+    { displayName
+    , initialState
     , receiveProps: mkEffFn3 receiveProps
     , render: mkFn3 render
     }
 
-foreign import component_ :: forall props. Fn2 (ReactComponent props) props JSX
+-- | SetState uses an update function to modify the current state.
+type SetState state fx = (state -> state) -> Eff (react :: ReactFX | fx) Unit
```

Comment:
👍 for moving to `state -> state`. I think this could let us build more interesting abstraction on top, without adding dependencies and complexity here.

Reply:
This is a good change. I was playing with this library this weekend and add state issues with the current approach because of react and batching. This solved the issue I ran into.